singletons-2.5 regression: derived Show instances for recursive singleton types loop at compile time

Unfortunately, I inadvertently introduced a regression in the way derived `Show` instances for recursive singleton types work. To explain what I mean, consider the following code:

```haskell
{-# LANGUAGE CPP #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE UndecidableInstances #-}

#if __GLASGOW_HASKELL__ >= 806
{-# LANGUAGE QuantifiedConstraints #-}
#endif
module Bug where

import Data.Kind
import Data.Singletons.TH

$(singletons [d|
  data X (a :: Type) = X1 | X2 (Y a) deriving Show
  data Y (a :: Type) = Y1 | Y2 (X a) deriving Show
  |])

main :: IO ()
main = print (sing :: Sing (X1 :: X Bool))
```

This compiles without issue in `singletons-2.4`. In `singletons-2.5`, however, it fails to compile with the following error:

```
$ /opt/ghc/8.6.2/bin/ghc Bug.hs
[1 of 1] Compiling Bug              ( Bug.hs, Bug.o )

Bug.hs:27:8: error:
    • Reduction stack overflow; size = 201
      When simplifying the following type: Show (Sing z)
      Use -freduction-depth=0 to disable this check
      (any upper bound you could choose might fail unpredictably with
       minor updates to GHC, so disabling the check is recommended if
       you're sure that type checking should terminate)
    • In the expression: print (sing :: Sing (X1 :: X Bool))
      In an equation for ‘main’:
          main = print (sing :: Sing (X1 :: X Bool))
   |
27 | main = print (sing :: Sing (X1 :: X Bool))
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

To see why this happens, consider the code that gets generated for the derived `Show` instances for `X` and `Y`:

```haskell
deriving instance ShowSing (Y a) => Show (Sing (z :: X a))
deriving instance ShowSing (X a) => Show (Sing (z :: Y a))
```

This part typechecks fine. It's when you actually try to _use_ it, as in `print (sing :: Sing (X1 :: X Bool))`, where the typechecker loops forever. Recall that `ShowSing` was changed in `singletons-2.5` to be a type synonym:

```haskell
type ShowSing k = (forall (z :: k). Show (Sing z) :: Constraint)
```

If you expand both occurrences of the `ShowSing` type synonym in the generated instances, you'll get:

```haskell
deriving instance (forall z. Show (Sing (z :: Y a))) => Show (Sing z :: X a)
deriving instance (forall z. Show (Sing (z :: X a))) => Show (Sing z :: Y a)
```

Due to the way `QuantifiedConstraints` works, GHC will always favor local, quantified constraints in the instance contexts over top-level instances. Notice that there is both a top-level instance for `Y a` as well as a local, quantified instance for `Y a` in scope in the `Show` instance for `X a`, so GHC will favor the local instance during instance resolution. But when resolving the local instance for `Y a`, we are back in the same situation: there is both a top-level and local instance for `X a`, so the local instance is picked. When resolving _that_ instance... we repeat the process and find ourselves in an infinite loop that eventually overflows the reduction stack. Eep.

I think the simplest solution to make `ShowSing` a class again. That is, redefine it like this:

```haskell
class    (forall (z :: k). Show (Sing z)) => ShowSing k
instance (forall (z :: k). Show (Sing z)) => ShowSing k
```

All existing `singletons` code should continue to work with this version of `ShowSing`, and since using a class ties the recursive knot, it doesn't suffer from the infinite looping issue that is described above. As an added bonus, users no longer have to enable `QuantifiedConstraints` in all every module that singles a derived `Show` instance, since the use of `QuantifiedConstraints` is now localized entirely to the module that defines the `ShowSing` class.

Does this sound reasonable, @goldfirere? If so, this might be worth releasing a `singletons-2.5.1` point release for, since this is a pretty nasty regression that prevents things like `show (sing :: Sing '[True])` from typechecking, as I recently discovered.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -5,6 +5,14 @@ next
 ----
 * Add an `IsString` instance for `SomeSing`.
 
+2.5.1
+-----
+* `ShowSing` is now a type class (with a single instance) instead of a type
+  synonym. This was changed because defining `ShowSing` as a type synonym
+  prevents it from working well with recursive types.
+  For more information, see
+  [issue #371](https://github.com/goldfirere/singletons/issues/371).
+
 2.5
 ---
 * The `Data.Promotion.Prelude.*` namespace has been removed. Use the

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -9,8 +9,8 @@ next
 -----
 * `ShowSing` is now a type class (with a single instance) instead of a type
   synonym. This was changed because defining `ShowSing` as a type synonym
-  prevents it from working well with recursive types.
-  For more information, see
+  prevents it from working well with recursive types due to an unfortunate GHC
+  bug. For more information, see
   [issue #371](https://github.com/goldfirere/singletons/issues/371).
 
 2.5

--- a/README.md
+++ b/README.md
@@ -56,7 +56,6 @@ following:
 * `KindSignatures`
 * `NoStarIsType`
 * `PolyKinds`
-* `QuantifiedConstraints`
 * `RankNTypes`
 * `ScopedTypeVariables`
 * `StandaloneDeriving`

--- a/src/Data/Singletons.hs
+++ b/src/Data/Singletons.hs
@@ -5,7 +5,6 @@
 {-# LANGUAGE MagicHash #-}
 {-# LANGUAGE PatternSynonyms #-}
 {-# LANGUAGE PolyKinds #-}
-{-# LANGUAGE QuantifiedConstraints #-}
 {-# LANGUAGE StandaloneDeriving #-}
 {-# LANGUAGE TemplateHaskell #-}
 {-# LANGUAGE TypeFamilies #-}

--- a/src/Data/Singletons/Prelude/Monoid.hs
+++ b/src/Data/Singletons/Prelude/Monoid.hs
@@ -4,7 +4,6 @@
 {-# LANGUAGE GADTs #-}
 {-# LANGUAGE InstanceSigs #-}
 {-# LANGUAGE PolyKinds #-}
-{-# LANGUAGE QuantifiedConstraints #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE StandaloneDeriving #-}
 {-# LANGUAGE TemplateHaskell #-}

--- a/src/Data/Singletons/Prelude/Semigroup.hs
+++ b/src/Data/Singletons/Prelude/Semigroup.hs
@@ -4,7 +4,6 @@
 {-# LANGUAGE GADTs #-}
 {-# LANGUAGE InstanceSigs #-}
 {-# LANGUAGE PolyKinds #-}
-{-# LANGUAGE QuantifiedConstraints #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE StandaloneDeriving #-}
 {-# LANGUAGE TemplateHaskell #-}

--- a/src/Data/Singletons/ShowSing.hs
+++ b/src/Data/Singletons/ShowSing.hs
@@ -29,7 +29,6 @@ module Data.Singletons.ShowSing (
   ShowSing
   ) where
 
-import Data.Kind
 import Data.Singletons.Internal
 import Data.Singletons.Prelude.Instances
 import Data.Singletons.Single
@@ -70,7 +69,7 @@ import qualified GHC.TypeNats as TN
 -- @
 --
 -- Because that quantified constraint is somewhat lengthy, we provide the
--- 'ShowSing' type synonym as a convenient shorthand. Thus, the above instance
+-- 'ShowSing' class synonym as a convenient shorthand. Thus, the above instance
 -- is equivalent to:
 --
 -- @
@@ -87,7 +86,78 @@ import qualified GHC.TypeNats as TN
 -- * A 'Show' instance for the singleton type
 --
 -- What a bargain!
-type ShowSing k = (forall z. Show (Sing (z :: k)) :: Constraint)
+class    (forall (z :: k). Show (Sing z)) => ShowSing k
+instance (forall (z :: k). Show (Sing z)) => ShowSing k
+
+{-
+Note [ShowSing must be a class, not a type synonym]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+It is very, very tempting to define ShowSing as a type synonym, like this:
+
+  type ShowSing k = (forall (z :: k). Show (Sing z) :: Constraint)
+
+In fact, version 2.5 of the library made the mistake of doing this. However,
+we have since learned the folly of our ways and now define ShowSing as class
+with a single instance.
+
+Why shouldn't you define ShowSing as a type synonym? Recursive data types.
+Consider the following example (from #371):
+
+  $(singletons [d|
+    data X a = X1 | X2 (Y a) deriving Show
+    data Y a = Y1 | Y2 (X a) deriving Show
+    |])
+
+This will generate the following instances:
+
+  deriving instance ShowSing (Y a) => Show (Sing (z :: X a))
+  deriving instance ShowSing (X a) => Show (Sing (z :: Y a))
+
+So far, so good. Now, suppose you try to actually `show` a singleton for X.
+For example:
+
+  show (sing @(X1 :: X Bool))
+
+Somewhat surprisingly, this will be rejected by the typechecker with the
+following error:
+
+    • Reduction stack overflow; size = 201
+      When simplifying the following type: Show (Sing z)
+
+To see why this happens, observe what goes on if we expand the occurrences of
+the ShowSing type synonym in the generated instances:
+
+  deriving instance (forall z. Show (Sing (z :: Y a))) => Show (Sing z :: X a)
+  deriving instance (forall z. Show (Sing (z :: X a))) => Show (Sing z :: Y a)
+
+Due to the way QuantifiedConstraints works, GHC will always favor local,
+quantified constraints in the instance contexts over top-level instances.
+Notice that there is both a top-level instance for (Y a) as well as a local,
+quantified instance for (Y a) in scope in the Show instance for (X a), so GHC
+will favor the local instance during instance resolution. But when resolving
+the local instance for (Y a), we are back in the same situation: there is both
+a top-level and local instance for (X a), so the local instance is picked.
+When resolving /that/ instance... we repeat the process and find ourselves in
+an infinite loop that eventually overflows the reduction stack. Eep.
+
+There are two possible ways to solve this problem:
+
+1. Make derived instances' type inference more clever. If you look closely,
+   you'll notice that the `ShowSing (X a)`/`ShowSing (Y a)` constraints in
+   the generated instances are entirely redundant and could safely be left
+   off. But determining this would require significantly improving singletons'
+   Template Haskell capabilities for type inference, which is a path that we
+   usually spurn in favor of keeping the generated code dumb but predictable.
+2. Define `ShowSing` as a class (with a single instance) instead of a type
+   synonym. `ShowSing`-as-a-class ties the recursive knot during instance
+   resolution and thus avoids the problems that the type synonym version
+   suffers from. As an added bonus, it also avoids the need for clients to
+   enable `QuantifiedConstraints` in every module that singles a derived
+   `Show` instance.
+
+Given the two options, (2) is by far the easier option, so that is what we
+ultimately went with.
+-}
 
 ------------------------------------------------------------
 -- TypeLits instances

--- a/src/Data/Singletons/ShowSing.hs
+++ b/src/Data/Singletons/ShowSing.hs
@@ -90,17 +90,20 @@ class    (forall (z :: k). Show (Sing z)) => ShowSing k
 instance (forall (z :: k). Show (Sing z)) => ShowSing k
 
 {-
-Note [ShowSing must be a class, not a type synonym]
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-It is very, very tempting to define ShowSing as a type synonym, like this:
+Note [Define ShowSing as a class, not a type synonym]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+In an ideal world, we would simply define ShowSing as an ordinary type synonym,
+like this:
 
   type ShowSing k = (forall (z :: k). Show (Sing z) :: Constraint)
 
-In fact, version 2.5 of the library made the mistake of doing this. However,
-we have since learned the folly of our ways and now define ShowSing as class
-with a single instance.
+In fact, I used to assume that we lived in an ideal world, so I defined
+ShowSing as a type synonym in version 2.5 of this library. However, I realized
+some time after 2.5's release that the world is far from ideal, unfortunately,
+and that this approach is unfeasible at the time being due to GHC Trac #15888.
 
-Why shouldn't you define ShowSing as a type synonym? Recursive data types.
+To be more precise, the exact issue involves an infelicity in the way
+QuantifiedConstraints interacts with recursive type class instances.
 Consider the following example (from #371):
 
   $(singletons [d|
@@ -127,20 +130,23 @@ following error:
 To see why this happens, observe what goes on if we expand the occurrences of
 the ShowSing type synonym in the generated instances:
 
-  deriving instance (forall z. Show (Sing (z :: Y a))) => Show (Sing z :: X a)
-  deriving instance (forall z. Show (Sing (z :: X a))) => Show (Sing z :: Y a)
+  deriving instance (forall z. Show (Sing (z :: Y a))) => Show (Sing (z :: X a))
+  deriving instance (forall z. Show (Sing (z :: X a))) => Show (Sing (z :: Y a))
 
-Due to the way QuantifiedConstraints works, GHC will always favor local,
-quantified constraints in the instance contexts over top-level instances.
-Notice that there is both a top-level instance for (Y a) as well as a local,
-quantified instance for (Y a) in scope in the Show instance for (X a), so GHC
-will favor the local instance during instance resolution. But when resolving
-the local instance for (Y a), we are back in the same situation: there is both
-a top-level and local instance for (X a), so the local instance is picked.
-When resolving /that/ instance... we repeat the process and find ourselves in
-an infinite loop that eventually overflows the reduction stack. Eep.
+Due to the way QuantifiedConstraints currently works (as surmised in Trac
+#15888), when GHC has a Wanted `Show (Sing X1 :: X Bool)` constraint, it
+chooses the appropriate instance and emits a Wanted
+`forall z. Show (Sing (z :: Y Bool))` constraint (from the instance context).
+GHC skolemizes the `z` to `z1` and tries to solve a Wanted
+`Show (Sing (z1 :: Y Bool))` constraint. GHC chooses the appropriate instance
+and emits a Wanted `forall z. Show (Sing (z :: X Bool))` constraint. GHC
+skolemizes the `z` to `z2` and tries to solve a Wanted
+`Show (Sing (z2 :: X Bool))` constraint... we repeat the process and find
+ourselves in an infinite loop that eventually overflows the reduction stack.
+Eep.
 
-There are two possible ways to solve this problem:
+Until Trac #15888 is fixed, there are two possible ways to work around this
+problem:
 
 1. Make derived instances' type inference more clever. If you look closely,
    you'll notice that the `ShowSing (X a)`/`ShowSing (Y a)` constraints in
@@ -151,9 +157,7 @@ There are two possible ways to solve this problem:
 2. Define `ShowSing` as a class (with a single instance) instead of a type
    synonym. `ShowSing`-as-a-class ties the recursive knot during instance
    resolution and thus avoids the problems that the type synonym version
-   suffers from. As an added bonus, it also avoids the need for clients to
-   enable `QuantifiedConstraints` in every module that singles a derived
-   `Show` instance.
+   currently suffers from.
 
 Given the two options, (2) is by far the easier option, so that is what we
 ultimately went with.

--- a/tests/SingletonsTestSuite.hs
+++ b/tests/SingletonsTestSuite.hs
@@ -102,6 +102,7 @@ tests =
     , compileAndDumpStdTest "FunctorLikeDeriving"
     , compileAndDumpStdTest "T353"
     , compileAndDumpStdTest "T358"
+    , compileAndDumpStdTest "T371"
     ],
     testCompileAndDumpGroup "Promote"
     [ compileAndDumpStdTest "Constructors"

--- a/tests/SingletonsTestSuiteUtils.hs
+++ b/tests/SingletonsTestSuiteUtils.hs
@@ -69,7 +69,6 @@ ghcOpts = ghcFlags ++ [
   , "-XTypeApplications"
   , "-XEmptyCase"
   , "-XNoStarIsType"
-  , "-XQuantifiedConstraints"
   ]
 
 -- Compile a test using specified GHC options. Save output to file, filter with

--- a/tests/compile-and-dump/Singletons/T371.ghc86.template
+++ b/tests/compile-and-dump/Singletons/T371.ghc86.template
@@ -1,0 +1,233 @@
+Singletons/T371.hs:(0,0)-(0,0): Splicing declarations
+    singletons
+      [d| data Y (a :: Type)
+            = Y1 | Y2 (X a)
+            deriving Show
+          data X (a :: Type)
+            = X1 | X2 (Y a)
+            deriving Show |]
+  ======>
+    data X (a :: Type)
+      = X1 | X2 (Y a)
+      deriving Show
+    data Y (a :: Type)
+      = Y1 | Y2 (X a)
+      deriving Show
+    type X1Sym0 = X1
+    type X2Sym1 (t0123456789876543210 :: Y a0123456789876543210) =
+        X2 t0123456789876543210
+    instance SuppressUnusedWarnings X2Sym0 where
+      suppressUnusedWarnings = snd (((,) X2Sym0KindInference) ())
+    data X2Sym0 :: forall (a0123456789876543210 :: Type).
+                   (~>) (Y a0123456789876543210) (X (a0123456789876543210 :: Type))
+      where
+        X2Sym0KindInference :: forall t0123456789876543210
+                                      arg. SameKind (Apply X2Sym0 arg) (X2Sym1 arg) =>
+                               X2Sym0 t0123456789876543210
+    type instance Apply X2Sym0 t0123456789876543210 = X2 t0123456789876543210
+    type Y1Sym0 = Y1
+    type Y2Sym1 (t0123456789876543210 :: X a0123456789876543210) =
+        Y2 t0123456789876543210
+    instance SuppressUnusedWarnings Y2Sym0 where
+      suppressUnusedWarnings = snd (((,) Y2Sym0KindInference) ())
+    data Y2Sym0 :: forall (a0123456789876543210 :: Type).
+                   (~>) (X a0123456789876543210) (Y (a0123456789876543210 :: Type))
+      where
+        Y2Sym0KindInference :: forall t0123456789876543210
+                                      arg. SameKind (Apply Y2Sym0 arg) (Y2Sym1 arg) =>
+                               Y2Sym0 t0123456789876543210
+    type instance Apply Y2Sym0 t0123456789876543210 = Y2 t0123456789876543210
+    type family ShowsPrec_0123456789876543210 (a :: GHC.Types.Nat) (a :: X a) (a :: GHC.Types.Symbol) :: GHC.Types.Symbol where
+      ShowsPrec_0123456789876543210 _ X1 a_0123456789876543210 = Apply (Apply ShowStringSym0 "X1") a_0123456789876543210
+      ShowsPrec_0123456789876543210 p_0123456789876543210 (X2 arg_0123456789876543210) a_0123456789876543210 = Apply (Apply (Apply ShowParenSym0 (Apply (Apply (>@#@$) p_0123456789876543210) (Data.Singletons.Prelude.Num.FromInteger 10))) (Apply (Apply (.@#@$) (Apply ShowStringSym0 "X2 ")) (Apply (Apply ShowsPrecSym0 (Data.Singletons.Prelude.Num.FromInteger 11)) arg_0123456789876543210))) a_0123456789876543210
+    type ShowsPrec_0123456789876543210Sym3 (a0123456789876543210 :: GHC.Types.Nat) (a0123456789876543210 :: X a0123456789876543210) (a0123456789876543210 :: GHC.Types.Symbol) =
+        ShowsPrec_0123456789876543210 a0123456789876543210 a0123456789876543210 a0123456789876543210
+    instance SuppressUnusedWarnings (ShowsPrec_0123456789876543210Sym2 a0123456789876543210 a0123456789876543210) where
+      suppressUnusedWarnings
+        = snd (((,) ShowsPrec_0123456789876543210Sym2KindInference) ())
+    data ShowsPrec_0123456789876543210Sym2 (a0123456789876543210 :: GHC.Types.Nat) (a0123456789876543210 :: X a0123456789876543210) :: (~>) GHC.Types.Symbol GHC.Types.Symbol
+      where
+        ShowsPrec_0123456789876543210Sym2KindInference :: forall a0123456789876543210
+                                                                 a0123456789876543210
+                                                                 a0123456789876543210
+                                                                 arg. SameKind (Apply (ShowsPrec_0123456789876543210Sym2 a0123456789876543210 a0123456789876543210) arg) (ShowsPrec_0123456789876543210Sym3 a0123456789876543210 a0123456789876543210 arg) =>
+                                                          ShowsPrec_0123456789876543210Sym2 a0123456789876543210 a0123456789876543210 a0123456789876543210
+    type instance Apply (ShowsPrec_0123456789876543210Sym2 a0123456789876543210 a0123456789876543210) a0123456789876543210 = ShowsPrec_0123456789876543210 a0123456789876543210 a0123456789876543210 a0123456789876543210
+    instance SuppressUnusedWarnings (ShowsPrec_0123456789876543210Sym1 a0123456789876543210) where
+      suppressUnusedWarnings
+        = snd (((,) ShowsPrec_0123456789876543210Sym1KindInference) ())
+    data ShowsPrec_0123456789876543210Sym1 (a0123456789876543210 :: GHC.Types.Nat) :: forall a0123456789876543210.
+                                                                                      (~>) (X a0123456789876543210) ((~>) GHC.Types.Symbol GHC.Types.Symbol)
+      where
+        ShowsPrec_0123456789876543210Sym1KindInference :: forall a0123456789876543210
+                                                                 a0123456789876543210
+                                                                 arg. SameKind (Apply (ShowsPrec_0123456789876543210Sym1 a0123456789876543210) arg) (ShowsPrec_0123456789876543210Sym2 a0123456789876543210 arg) =>
+                                                          ShowsPrec_0123456789876543210Sym1 a0123456789876543210 a0123456789876543210
+    type instance Apply (ShowsPrec_0123456789876543210Sym1 a0123456789876543210) a0123456789876543210 = ShowsPrec_0123456789876543210Sym2 a0123456789876543210 a0123456789876543210
+    instance SuppressUnusedWarnings ShowsPrec_0123456789876543210Sym0 where
+      suppressUnusedWarnings
+        = snd (((,) ShowsPrec_0123456789876543210Sym0KindInference) ())
+    data ShowsPrec_0123456789876543210Sym0 :: forall a0123456789876543210.
+                                              (~>) GHC.Types.Nat ((~>) (X a0123456789876543210) ((~>) GHC.Types.Symbol GHC.Types.Symbol))
+      where
+        ShowsPrec_0123456789876543210Sym0KindInference :: forall a0123456789876543210
+                                                                 arg. SameKind (Apply ShowsPrec_0123456789876543210Sym0 arg) (ShowsPrec_0123456789876543210Sym1 arg) =>
+                                                          ShowsPrec_0123456789876543210Sym0 a0123456789876543210
+    type instance Apply ShowsPrec_0123456789876543210Sym0 a0123456789876543210 = ShowsPrec_0123456789876543210Sym1 a0123456789876543210
+    instance PShow (X a) where
+      type ShowsPrec a a a = Apply (Apply (Apply ShowsPrec_0123456789876543210Sym0 a) a) a
+    type family ShowsPrec_0123456789876543210 (a :: GHC.Types.Nat) (a :: Y a) (a :: GHC.Types.Symbol) :: GHC.Types.Symbol where
+      ShowsPrec_0123456789876543210 _ Y1 a_0123456789876543210 = Apply (Apply ShowStringSym0 "Y1") a_0123456789876543210
+      ShowsPrec_0123456789876543210 p_0123456789876543210 (Y2 arg_0123456789876543210) a_0123456789876543210 = Apply (Apply (Apply ShowParenSym0 (Apply (Apply (>@#@$) p_0123456789876543210) (Data.Singletons.Prelude.Num.FromInteger 10))) (Apply (Apply (.@#@$) (Apply ShowStringSym0 "Y2 ")) (Apply (Apply ShowsPrecSym0 (Data.Singletons.Prelude.Num.FromInteger 11)) arg_0123456789876543210))) a_0123456789876543210
+    type ShowsPrec_0123456789876543210Sym3 (a0123456789876543210 :: GHC.Types.Nat) (a0123456789876543210 :: Y a0123456789876543210) (a0123456789876543210 :: GHC.Types.Symbol) =
+        ShowsPrec_0123456789876543210 a0123456789876543210 a0123456789876543210 a0123456789876543210
+    instance SuppressUnusedWarnings (ShowsPrec_0123456789876543210Sym2 a0123456789876543210 a0123456789876543210) where
+      suppressUnusedWarnings
+        = snd (((,) ShowsPrec_0123456789876543210Sym2KindInference) ())
+    data ShowsPrec_0123456789876543210Sym2 (a0123456789876543210 :: GHC.Types.Nat) (a0123456789876543210 :: Y a0123456789876543210) :: (~>) GHC.Types.Symbol GHC.Types.Symbol
+      where
+        ShowsPrec_0123456789876543210Sym2KindInference :: forall a0123456789876543210
+                                                                 a0123456789876543210
+                                                                 a0123456789876543210
+                                                                 arg. SameKind (Apply (ShowsPrec_0123456789876543210Sym2 a0123456789876543210 a0123456789876543210) arg) (ShowsPrec_0123456789876543210Sym3 a0123456789876543210 a0123456789876543210 arg) =>
+                                                          ShowsPrec_0123456789876543210Sym2 a0123456789876543210 a0123456789876543210 a0123456789876543210
+    type instance Apply (ShowsPrec_0123456789876543210Sym2 a0123456789876543210 a0123456789876543210) a0123456789876543210 = ShowsPrec_0123456789876543210 a0123456789876543210 a0123456789876543210 a0123456789876543210
+    instance SuppressUnusedWarnings (ShowsPrec_0123456789876543210Sym1 a0123456789876543210) where
+      suppressUnusedWarnings
+        = snd (((,) ShowsPrec_0123456789876543210Sym1KindInference) ())
+    data ShowsPrec_0123456789876543210Sym1 (a0123456789876543210 :: GHC.Types.Nat) :: forall a0123456789876543210.
+                                                                                      (~>) (Y a0123456789876543210) ((~>) GHC.Types.Symbol GHC.Types.Symbol)
+      where
+        ShowsPrec_0123456789876543210Sym1KindInference :: forall a0123456789876543210
+                                                                 a0123456789876543210
+                                                                 arg. SameKind (Apply (ShowsPrec_0123456789876543210Sym1 a0123456789876543210) arg) (ShowsPrec_0123456789876543210Sym2 a0123456789876543210 arg) =>
+                                                          ShowsPrec_0123456789876543210Sym1 a0123456789876543210 a0123456789876543210
+    type instance Apply (ShowsPrec_0123456789876543210Sym1 a0123456789876543210) a0123456789876543210 = ShowsPrec_0123456789876543210Sym2 a0123456789876543210 a0123456789876543210
+    instance SuppressUnusedWarnings ShowsPrec_0123456789876543210Sym0 where
+      suppressUnusedWarnings
+        = snd (((,) ShowsPrec_0123456789876543210Sym0KindInference) ())
+    data ShowsPrec_0123456789876543210Sym0 :: forall a0123456789876543210.
+                                              (~>) GHC.Types.Nat ((~>) (Y a0123456789876543210) ((~>) GHC.Types.Symbol GHC.Types.Symbol))
+      where
+        ShowsPrec_0123456789876543210Sym0KindInference :: forall a0123456789876543210
+                                                                 arg. SameKind (Apply ShowsPrec_0123456789876543210Sym0 arg) (ShowsPrec_0123456789876543210Sym1 arg) =>
+                                                          ShowsPrec_0123456789876543210Sym0 a0123456789876543210
+    type instance Apply ShowsPrec_0123456789876543210Sym0 a0123456789876543210 = ShowsPrec_0123456789876543210Sym1 a0123456789876543210
+    instance PShow (Y a) where
+      type ShowsPrec a a a = Apply (Apply (Apply ShowsPrec_0123456789876543210Sym0 a) a) a
+    data instance Sing :: X a -> Type
+      where
+        SX1 :: Sing X1
+        SX2 :: forall a (n :: Y a). (Sing (n :: Y a)) -> Sing (X2 n)
+    type SX = (Sing :: X a -> Type)
+    instance SingKind a => SingKind (X a) where
+      type Demote (X a) = X (Demote a)
+      fromSing SX1 = X1
+      fromSing (SX2 b) = X2 (fromSing b)
+      toSing X1 = SomeSing SX1
+      toSing (X2 (b :: Demote (Y a)))
+        = case toSing b :: SomeSing (Y a) of {
+            SomeSing c -> SomeSing (SX2 c) }
+    data instance Sing :: Y a -> Type
+      where
+        SY1 :: Sing Y1
+        SY2 :: forall a (n :: X a). (Sing (n :: X a)) -> Sing (Y2 n)
+    type SY = (Sing :: Y a -> Type)
+    instance SingKind a => SingKind (Y a) where
+      type Demote (Y a) = Y (Demote a)
+      fromSing SY1 = Y1
+      fromSing (SY2 b) = Y2 (fromSing b)
+      toSing Y1 = SomeSing SY1
+      toSing (Y2 (b :: Demote (X a)))
+        = case toSing b :: SomeSing (X a) of {
+            SomeSing c -> SomeSing (SY2 c) }
+    instance SShow (Y a) => SShow (X a) where
+      sShowsPrec ::
+        forall (t1 :: GHC.Types.Nat) (t2 :: X a) (t3 :: GHC.Types.Symbol).
+        Sing t1
+        -> Sing t2
+           -> Sing t3
+              -> Sing (Apply (Apply (Apply (ShowsPrecSym0 :: TyFun GHC.Types.Nat ((~>) (X a) ((~>) GHC.Types.Symbol GHC.Types.Symbol))
+                                                             -> Type) t1) t2) t3)
+      sShowsPrec
+        _
+        SX1
+        (sA_0123456789876543210 :: Sing a_0123456789876543210)
+        = (applySing
+             ((applySing ((singFun2 @ShowStringSym0) sShowString))
+                (sing :: Sing "X1")))
+            sA_0123456789876543210
+      sShowsPrec
+        (sP_0123456789876543210 :: Sing p_0123456789876543210)
+        (SX2 (sArg_0123456789876543210 :: Sing arg_0123456789876543210))
+        (sA_0123456789876543210 :: Sing a_0123456789876543210)
+        = (applySing
+             ((applySing
+                 ((applySing ((singFun3 @ShowParenSym0) sShowParen))
+                    ((applySing
+                        ((applySing ((singFun2 @(>@#@$)) (%>))) sP_0123456789876543210))
+                       (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 10)))))
+                ((applySing
+                    ((applySing ((singFun3 @(.@#@$)) (%.)))
+                       ((applySing ((singFun2 @ShowStringSym0) sShowString))
+                          (sing :: Sing "X2 "))))
+                   ((applySing
+                       ((applySing ((singFun3 @ShowsPrecSym0) sShowsPrec))
+                          (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 11))))
+                      sArg_0123456789876543210))))
+            sA_0123456789876543210
+    instance SShow (X a) => SShow (Y a) where
+      sShowsPrec ::
+        forall (t1 :: GHC.Types.Nat) (t2 :: Y a) (t3 :: GHC.Types.Symbol).
+        Sing t1
+        -> Sing t2
+           -> Sing t3
+              -> Sing (Apply (Apply (Apply (ShowsPrecSym0 :: TyFun GHC.Types.Nat ((~>) (Y a) ((~>) GHC.Types.Symbol GHC.Types.Symbol))
+                                                             -> Type) t1) t2) t3)
+      sShowsPrec
+        _
+        SY1
+        (sA_0123456789876543210 :: Sing a_0123456789876543210)
+        = (applySing
+             ((applySing ((singFun2 @ShowStringSym0) sShowString))
+                (sing :: Sing "Y1")))
+            sA_0123456789876543210
+      sShowsPrec
+        (sP_0123456789876543210 :: Sing p_0123456789876543210)
+        (SY2 (sArg_0123456789876543210 :: Sing arg_0123456789876543210))
+        (sA_0123456789876543210 :: Sing a_0123456789876543210)
+        = (applySing
+             ((applySing
+                 ((applySing ((singFun3 @ShowParenSym0) sShowParen))
+                    ((applySing
+                        ((applySing ((singFun2 @(>@#@$)) (%>))) sP_0123456789876543210))
+                       (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 10)))))
+                ((applySing
+                    ((applySing ((singFun3 @(.@#@$)) (%.)))
+                       ((applySing ((singFun2 @ShowStringSym0) sShowString))
+                          (sing :: Sing "Y2 "))))
+                   ((applySing
+                       ((applySing ((singFun3 @ShowsPrecSym0) sShowsPrec))
+                          (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 11))))
+                      sArg_0123456789876543210))))
+            sA_0123456789876543210
+    deriving instance Data.Singletons.ShowSing.ShowSing (Y a) =>
+                      Show (Sing (z :: X a))
+    deriving instance Data.Singletons.ShowSing.ShowSing (X a) =>
+                      Show (Sing (z :: Y a))
+    instance SingI X1 where
+      sing = SX1
+    instance SingI n => SingI (X2 (n :: Y a)) where
+      sing = SX2 sing
+    instance SingI (X2Sym0 :: (~>) (Y a) (X (a :: Type))) where
+      sing = (singFun1 @X2Sym0) SX2
+    instance SingI (TyCon1 X2 :: (~>) (Y a) (X (a :: Type))) where
+      sing = (singFun1 @(TyCon1 X2)) SX2
+    instance SingI Y1 where
+      sing = SY1
+    instance SingI n => SingI (Y2 (n :: X a)) where
+      sing = SY2 sing
+    instance SingI (Y2Sym0 :: (~>) (X a) (Y (a :: Type))) where
+      sing = (singFun1 @Y2Sym0) SY2
+    instance SingI (TyCon1 Y2 :: (~>) (X a) (Y (a :: Type))) where
+      sing = (singFun1 @(TyCon1 Y2)) SY2

--- a/tests/compile-and-dump/Singletons/T371.hs
+++ b/tests/compile-and-dump/Singletons/T371.hs
@@ -1,0 +1,16 @@
+module T371 where
+
+import Data.Kind
+import Data.Singletons.TH
+
+$(singletons [d|
+  data X (a :: Type) = X1 | X2 (Y a) deriving Show
+  data Y (a :: Type) = Y1 | Y2 (X a) deriving Show
+  |])
+
+main :: IO ()
+main = do
+  print (sing :: Sing ('[] :: [Bool]))
+  print (sing :: Sing '[True])
+  print (sing :: Sing (X1 :: X Bool))
+  print (sing :: Sing (Y2 X1 :: Y Bool))

